garrot02/garrot.1
.TH GARROT 1 "07 Dec 1994" "Linux" "DOSEMU User's Manual"
.SH NAME
garrot \- release DOSEMU idle-time back to system
.SH SYNOPSIS
.B garrot
[ -h | -DDD | -n DDD | -v | -q | -t ]
.nf
 \-h      help screen
 \-DDD    set garrot constant to DDD (an integer number)
 \-n DDD  alternate syntax for setting the garrot constant
 \-v      display version
 \-q      query current settings
 \-t      show terms of license
.fi
.SH DESCRIPTION
This manual page
documents 
.BR garrot, 
a support program for the Linux DOSEMU subsystem. NOTE: garrot
runs inside of the DOSEMU subsystem, 
.B not
from the Linux
command line.  Garrot improves Linux performance by coercing the
operating system running inside of DOSEMU to release the CPU back to
the main system during idle periods.  CPU usage as reported by 'top'
decreases dramatically, dropping down from 98% to 1% or less. The
current version of Garrot is 0.2, and it should be considered
pre-alpha software.

.PP
Garrot runs inside of the emulator session. You must move the program
garrot.com into the emulator. You may then run it from the "DOS" command
prompt, usually giving it a numeric argument dubbed "the garrot
constant", for example:
.nf

                    C:> garrot -12

.fi
Garrot installs itself as a daemon and returns you to the DOSEMU
command prompt.  You may re-run garrot at any time if you wish to
change the original argument.  
.PP
The garrot constant is the numeric argument which provides the optimal
balance between DOSEMU performance and idle-time cpu usage.  It appears
that this optimal value is approximately equal to one-half of your Linux
BogoMips value. For example, if your system reports a BogoMips value of
20, then choosing a garrot constant value of 10 will be a good first
approximation. Garrot uses this number as a counter threshold for
pre-empting DOSEMU after detecting a certain period of idleness. Lower
values imply less tolerance for idle time CPU use. 
.PP
You may determine this empirically by running 'top' in one virtual
console and running DOSEMU in another.  CPU usage as reported by 'top'
will drop to 2% or less at the critical value. It is a non-linear
function, and the change is abrupt. Choose successively lower values
for the garrot constant until you see the drop. Allow a couple moments
for 'top' to stabilize and update after changing the garrot value.
The garrot value which caused the sudden drop in CPU usage, or the value
one below it, is probably your optimal setting---the "garrot constant"
for your system. If in doubt, chose the value one below the drop point. 
.PP
Note that if you run Garrot without any command arguments, it will try
to guess an initial value for you. This value should be taken with a
grain of salt. Unfortunately the 
timers available inside of the DOSEMU sessions are not adequate to 
accurately calculate the garrot constant. Using one-half of your
BogoMips value is somewhat more accurate than the guess that garrot
can make. After you have determined the proper value, you may want
to add the garrot command to one of your "DOS" initialization files
(e.g. autoexec.bat) so that garrot will be invoked every time you
start a DOSEMU session. 
.PP
.SH BUGS
Some DOS programs that bypass certain parts of both the BIOS
and DOS system may cause CPU usage to rise because Garrot can not
figure out what it going on. Of course there may be blind spots in
garrot that need fixing. Please report any quirks to me, or
suggestions for improvement. Garrot has not been widely tested. It is
known to work with MS-DOS 5.0, DOSEMU 0.53pl28, Linux kernel 1.1.64.
Please forward your reports.
.PP
Garrot is offered "as-is"
without warranty of any kind (see the GPL). Even so, garrot does not
touch any risky part of the system. It doesn't involve disks or file
systems. 
.PP
.SH AUTHOR 
.PP
.nf
Thomas McWilliams:
  tgm@netcom.com
  thomas.mcwilliams@f615.n109.z1.fidonet.org
.fi
.SH COPYING
Distribution permitted under the terms of the GNU General Public
License, version 2 (or later versions at your preference). Copies
of this agreement are available from:
.nf

\	Free Software Foundation, Inc., 
\	675 Mass Ave, 
\	Cambridge, MA 02139, USA.

.fi
.SH "SEE ALSO"
.BR top "(1), " ps (1)


NEWS
NEWS file for DOSEMU
====================

This file contains a list of user visible changes in DOSEMU. For the
raw developer changelog please refer to ChangeLog.

Version 1.3.2 (2005-06-01)
=============
* Windows 3.1 now works natively, without any need for the WINOS2 patches.
* many other improvements to make Windows 3 in DOSEMU more stable and usable.
* fixed IF flag virtualization for DPMI, this makes many DPMI applications
  to work reliably without the random crashes (such as Transport Tycoon,
  Screamer, Vchess).
* Implemented LDT virtualization / direct LDT access for the DPMI clients.
* Many DPMI and PM API translator fixes to support all the commercial
  Borland tools and other protected-mode programs. You may need the linux
  kernel 2.6.12-rc3 or more recent versions however to get the truely
  reliable DPMI support.
* Most of the DPMI 1.0 API and functionality have been added.
* The redirected drive filesystem code is much faster than before.
* Fixed new errors and warnings with GCC 4.0.
* Allow a native himem.sys to be used in DOSEMU.
* Added the int15h PS/2 mouse interface.
* I/O ports are "fast" by default.
* Added some support for sound recording.
* the floppy is much faster with 2.6 kernels.
* added support for 2.88M floppies
* The printer code now uses line buffering so that you do not have to
  wait $_printer_timeout seconds before printing starts.
* TTY locking can be disabled by $_ttylocks=""
* IPX and Packet Driver code was seriously reworked.
* serial ports emulation was improved a lot in both reliability
  and performance.
* Terminal improvements:
  - UTF-8 terminals are supported if you use SLang 2 or a UTF-8
    patched SLang 1 library. The supplied SLang library subset is no
    longer necessary and no longer provided.
  - $_term_char_set no longer exists;
    $_external_char_set now defaults to the current locale.
  - Even on non UTF-8 terminals many line drawing characters can now
    be used, and on the Linux console almost the full CP437 character set.
  - Many ctrl, alt, shift combinations with function keys, arrow keys, etc,
    can now be used in xterm and friends.
  - VGA hardware is emulated in terminal mode as well as in X mode.
* Console improvements:
  - VESA driver added which should be compatible with most BIOSes. This
    is the default.
  - Linear frame buffers now work.
  - dosemu can now be started from under the FB console.
* X improvements:
  - Added extra X fonts (thanks to Martin Reuber)
  - The X aspect ratio is no longer set to 4:3 by default, but to an integer
    multiple of the real resolution.
  - much improved keyboard layout autodetection
  - fixed the grave bug which made the keyboard not to work properly in
    some games.
  - A graphical mouse cursor now works under X with mouse grab enabled.
  - Added keyboard grab - hardcoded to Ctrl-Alt-k for now.
* Added experimental SDL plugin. Mostly functional; makes it possible to run
  DOSEMU on all supported SDL backends.

List of SourceForge tracker numbers with bugs that were closed:
# 687024 parse /proc/pci for video ports
# 811365 Vchess crashe under dosemu 1.1.5.7 with mouse
# 818094 Mouse doesn't work in a Turbo Debugger
# 818145 certain keys doesn't work in xdosemu (i ,)
# 845119 (patch) 2.88 MB floppy support
# 855556 Problems running transport tycoon deluxe
# 855948 Problems with Screamer
# 990235 vgaemu crashes (regres)
# 991194 Int 21, AH=2F in protected mode doesnt work in all cases
# 999831 No upper case A or S 
#1004472 Problems with int 0x31, ah=01 functions
#1005015 Design error in DOSEMU's realmode callbacks?
#1007380 Compilation fails first time
#1012035 DOSEMU 1.3.1 : App hangs on COM Port access
#1019824 (patch) LFN support for DPMI host
#1019894 (patch) Fast port access
#1020635 bug in dosemu's LFN support
#1027805 Dos App hangs - PIC and DPMI related.
#1028236 About int 21, ax=0x4b00 in msdos.c
#1029819 serial port
#1033901 dosemu-CVS crashes DPMI
#1037915 Everything I type is double.
#1042020 serial port and cache register
#1067680 clipper take 99% of CPU
#1074310 Sound recording fails, playback is truncated
#1075068 dosemu 1.31. does not sense DCD
#1076593 (RFE) Windows support
#1079033 (RFE) mouse cursor in text mode with grab
#1080784 Compile Failure
#1081283 (SR) /proc access
#1089201 (patch) int15 block move support
#1103434 Serial communication fails with CVS version
#1110543 Custom keyboard layout
#1113468 Keyboard regression from v1.2 to latest CVS version
#1118031 (patch) fix target_cpu option
#1152829 dosemu should exit when the DOS process terminates
#1157059 (patch) command to automatically set sound env vars
#1164054 Saving fails in Settlers
#1178900 (SR) Joystick support!
#1189666 dbfcdx hangs on lredir drive

Version 1.3.1 (2004-07-11)
=============
* Support for 32rtm extender was added. It is now possible to use the
  32bit Borland tools under dosemu.
* Support for the HX DOS Extender was added. It is now possible to run
  some win32 console apps under dosemu, using that extender.
* The Blinker extender is now fully supported (all known versions).
  But some older versions of that extender may require the 2.4.x linux
  kernel (linux-2.6.8 will include the necessary fix).
* Improved DPMI uncommitted memory support (started in 1.3.0).
* The hogtheshold code was reorganized to be able to pass through certain
  events quicker (such as loading a file in FreeDOS EDIT) even when you 
  have it set to "1".
* Fixed all new warnings with GCC 3.4.
* Fixed compilation problem with Fedora Core 2 headers.
* Added cp895 and bg-mik character set support.
* Many VGAEMU bugs were fixed, and bitmap fonts are much quicker.
* Allow "fast" i/o ports above 0x400 for kernels that support it
  (2.6.7-mm kernels only for now)
* Implemented VxD hooks for WinOS2: seems to solve all remaining problems with
  Win3.1(WinOS2): allows Alt-Tab and the timer.
* Many DPMI and DOS extender fixes.
* An automatic backtrace using GDB (output in ~/.dosemu/boot.log)

garrot02/garrot.asm
;   
;   Garrot, a Linux dosemu support driver, releases idle time to system.
;   Copyright (C) 1994  Thomas G. McWilliams
;
;   This program is free software; you can redistribute it and/or modify
;   it under the terms of the GNU General Public License as published by
;   the Free Software Foundation; either version 2 of the License, or
;   (at your option) any later version.
;
;   This program is distributed in the hope that it will be useful,
;   but WITHOUT ANY WARRANTY; without even the implied warranty of
;   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;   GNU General Public License for more details.
;
;   You should have received a copy of the GNU General Public License
;   along with this program; if not, write to the Free Software
;   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
;
; author e-mail: tgm@netcom.com, thomas.mcwilliams@f615.n109.z1.fidonet.org
;

.286
TITLE garrot.asm 

PUBLIC	_begin,dvu,nice,flags,flagsend,i2dh_off,helpflg,versflg,queryflg
PUBLIC  termflg,runmsg,calmsg,lvlmsg,EOL,rezend,gvers
EXTRN _print:proc, _prnhx:proc, _getopt:proc, int_2dh:proc, _help:proc, _bogo:proc
EXTRN _gversion:proc, _terms:proc, _query:proc, _prneol:proc

DGROUP	GROUP	_TEXT
	ASSUME cs:DGROUP,ds:DGROUP,ss:DGROUP
_TEXT 	SEGMENT	WORD PUBLIC 'CODE'

KNICE equ 0ffffh
IVER  equ 2
SVER  equ IVER+32

i15h_vec:
        i15h_off	dw	0
        i15h_seg	dw	0

i16h_vec:
        i16hn_off	dw	0
        i16hn_seg	dw	0

i21h_vec:
        i21h_off	dw	0
        i21h_seg	dw	0

i28h_vec:
        i28h_off	dw      0
        i28h_seg	dw	0

i2dh_vec:
        i2dh_off	dw      0
        i2dh_seg	dw	0

vid_mode:
	vidmd_off	dw	449h
	vidmd_seg	dw	0h

vid_base:
	vidb_off	dw	0
	vidb_seg	dw	0b800h

flags:
	dvu 		db	0
	watchdog	db	0
	helpflg		db	0
	versflg		db	0
	queryflg	db	0
	termflg		db	0
	nice		dw	KNICE
flagsend:

int_28h_1 PROC FAR
	mov	al,cs:watchdog
	inc	al
	xor	ah,ah
	cmp	ax,cs:nice
	jb	nbv
        mov     ax,1680h 	 
        int     2fh 		
	xor	al,al
nbv:
	mov	cs:watchdog,al
	jmp dword ptr cs:i28h_vec   
int_28h_1 ENDP


int_16h_1 PROC FAR
	push 	es
	push	ax
	cmp	ah,0h	; trap problem call
	je	y1
	cmp	ah,10h	; trap problem call
	je	y1
	jmp	i16out
y1:
	call	chkchr	; poll for input
	jnz	chkct	; ok, key is ready, exit loop
	mov	ax,1680h  
	int	2fh	; release system
	jmp	y1

	 ;check count
chkct:	mov	al,cs:watchdog
	inc	al
i16out:	
	pop 	ax
	pop	es
	jmp dword ptr cs:i16h_vec
int_16h_1 ENDP



chkchr	PROC NEAR
	pushf
	push	cs
	push	offset c1
	mov	ah,11h		; see if keyboard is ready
	jmp dword ptr cs:i16h_vec
c1:	ret
chkchr	ENDP


int_15h_1 PROC FAR
        cmp     ax,1000h	; DV fake
	je	m1
	jmp dword ptr cs:i15h_vec   
m1:
        push    ax
        mov     ax,1680h 	; release system
        int     2fh 		
        pop     ax
        iret		
int_15h_1 ENDP

      
int_21h_ent   PROC    FAR
	cmp	ah,2ch		 ;check for call to time
	jne	i21out
	mov     ax,1680h 
        int     2fh                                 
	xor	al,al
	mov	cs:watchdog,al
	mov	ah,2ch
i21out:	jmp     dword ptr cs:i21h_vec  
int_21h_ent   ENDP

db   'Garrot, a Linux dosemu support driver, releases idle time to system',0 
db   'Copyright (C) 1994  Thomas G. McWilliams'
EOL	db	13,10,0

rezend:


newvect	PROC near
	push 	cs
	pop  	ds	
	 ;get video mode
	push	es 
	les	bx, dword ptr [vid_mode] 
	mov	al,7h
	cmp	es:[bx],al
	jne	lb1
	mov	ax,0b000h
	mov	vidb_seg,ax	
lb1:	pop	es

         ;get and save original int vector 15h 
        mov     ax,3515h
        int     21h                                
        mov     cs:i15h_off,bx	 ;store old int 15h offset 
        mov     cs:i15h_seg,es	 ;store old int 15h segment

	 ;set replacement  int 15h handler                           
        mov     dx,offset int_15h_1
        mov     ax,2515h 	
        int     21h  

	 ;save pointer to original int 16h handler
        mov     ax,3516h
        int     21h           
        mov     cs:i16hn_off,bx	 
        mov     cs:i16hn_seg,es	

	;set replacement  int 16h handler                           
        mov     dx,offset int_16h_1
        mov     ax,2516h 	
        int     21h  

	 ;save pointer to original int 21h handler
        mov     ax,3521h
        int     21h           
        mov     cs:i21h_off,bx	 ;store offset of old int 21h
        mov     cs:i21h_seg,es	 ;store segment of old int 21h

	 ;set replacement  int 21h handler                           
        mov     dx,offset int_21h_ent
        mov     ax,2521h 	
        int     21h  

	 ;save pointer to original int 28h handler
        mov     ax,3528h
        int     21h           
        mov     cs:i28h_off,bx	 
        mov     cs:i28h_seg,es	

	 ;set replacement  int 28h handler                           
        mov     dx,offset int_28h_1
        mov     ax,2528h 	
        int     21h  

	 ;save pointer to original int 2dh handler
        mov     ax,352dh
        int     21h           
        mov     cs:i2dh_off,bx	 
        mov     cs:i2dh_seg,es	

	 ;set replacement  int 2dh handler                           
        mov     dx,offset int_2dh
        mov     ax,252dh 	
        int     21h  

	ret
newvect ENDP 
        

notin:	db	'Garrot 0.2:  not currently installed',0
banner: db      'Garrot 0.2:  Linux DOSEMU support driver installed below ',0
runmsg  db	'Garrot 0.'
vmsg	db	'2:  running',0
errmsg: db	'Garrot 0.2:  error',0
calmsg:	db	'Garrot 0.2:  please wait, taking a guess . . . ',0
gvers	db	'garrot 0.2',0
lvlmsg: db	'level ',0

; below here is non-resident initialization code 

_begin: call	_getopt
	cmp	cx,0ffffh	; error code for _getopt
	je	err
	mov	al,helpflg
	cmp	al,0
	je	b1
	call	_help		; display help screen
	jmp	fin
b1:
	mov	al,versflg
	cmp	al,0
	je	b2
	call	_gversion
	jmp	fin
b2:
	mov	al,termflg
	cmp	al,0
	je	b3
	call	_terms
	jmp	fin
b3:
	mov     ax,0d300h	;check id, are we alredy installed?
        int     2dh        
        cmp     al,0ffh		 
        jne     doinit

	; already installed ...
	add	cl,30h	
	mov	vmsg,cl
	mov	bl,queryflg
	cmp	bl,0
	je	doupdate ; not a query, do update 

	call	stat
	cmp	al,0h
	je	fin	; exit ok
	jmp	fan	; exit error

doupdate:
	; update flags by passing pointer to int_2d 
	mov	ax,nice
	cmp	ax,KNICE
	jne	b8
	call	_bogo 	; try to guess a good value
b8:
	push	cs
	pop	dx
	mov	bx,offset flags
	mov	ax,0d306h	; update flags command
	int	2dh
	cmp	al,0h
	jne	err	; error

	call	stat	; no error
	jmp	short fin
err:
	push	offset errmsg
	call	_prneol	
	pop	ax

fan:	mov	ax,4c01h	 ;exit with error level 1
	jmp	short endo
fin:	mov	ax,4c00h	 ;exit with error level 0
endo:	int	21h	

doinit:	
	; check for query only
	mov	bl,queryflg
	cmp	bl,0h
	je	b6
	push	offset notin
	call	_prneol
	pop	ax
	jmp	fan	

b6:
	call newvect

        ;display title banner 
	push	offset banner
	call	_print
	pop	ax
	
	;print	location
	push	cs
	push offset rezend
	call	_prnhx
	pop	dx
	pop	dx
	mov	ax,nice
	cmp	ax,KNICE
	jne	b7
	call	_bogo 	; try to guess a good value
b7:	call	stat

	; terminate and stay resident
        mov     dx,offset rezend
	add	dx,15
	mov	cl,4h
	shr	dx,cl
	mov	ax,3100h
	int	21h

stat	PROC	near
	push	cx
	mov	ax,0d307h	; query flags
	int	2dh
	push	ax
	cmp	al,0h
	je	st1
	push	offset errmsg 
	call	_prneol	
	pop	ax	
	jmp	stend
st1:
	call	_query
stend:	pop	ax	
	pop	cx
	ret
stat	ENDP

_TEXT 	ENDS
	END 


